File: 백준/Silver/15654. N과 M （5）/N과 M （5）.cc
#include <bits/stdc++.h>

using namespace std;


int n, m;
int visited[10001];
vector<int> v;
vector<int> result;

void bt(int cnt)
{
	if (cnt == m)
	{
		for (int i = 0; i < m; i++)
		{
			cout << result[i] << " ";
		}
		cout << "\n";

		return;
	}

	for (int i = 0; i < n; i++)
	{
		if (!visited[i])
		{
			visited[i] = 1;
			result.push_back(v[i]);
			bt(cnt + 1);
			visited[i] = 0;
			result.pop_back();
		}
	}
}

int main()
{
	cin >> n >> m;

	for (int i = 0; i < n; i++)
	{
		int temp;
		cin >> temp;
		v.push_back(temp);
	}

	sort(v.begin(), v.end());

	bt(0);

	return 0;
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, m;
    static int[] result;
    static int[] arr;
    static boolean[] visited;

    public static void main(String[] args) throws Exception{
        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        arr = new int[n];

        st = new StringTokenizer(br.readLine());

        for (int i = 0 ; i < n ; i++){
            arr[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(arr);

        result = new int[m];
        visited = new boolean[n+1];

        dfs(0);

        System.out.println(sb);
    }

    static void dfs(int cur) {

        if (cur == m) {
            for (int i = 0; i < result.length ; i ++) {
                sb.append(result[i] + " ");
            }
            sb.append("\n");

            return;
        }

        for (int i = 0 ; i < n ; i ++) {
            if (!visited[i]) {
                visited[i] = true;
                result[cur] = arr[i];
                dfs(cur+1);
                visited[i] = false;
            }
        }
    }

}